osbuild/buildroot.py

```diff
@@ -164,6 +169,10 @@ def run(self, argv, binds=None, readonly_binds=None):
         for b in readonly_binds or []:
             mounts += ["--ro-bind"] + b.split(":")
 
+        # Prepare all registered API endpoints
+        for api in self._apis:
+            self._exitstack.enter_context(api)
+
```

Comment:
Why do you do that in `run()` rather than in `__enter__()`? Before this, you could re-use the buildroot for multiple runs, but with this you would suddenly get all the APIs entered multiple times, right?

Maybe this is required by the callers, because you already are in the `with` context when registering APIs. But in that case you could also "drain" `self._apis` when entering the context, thus calling `register_api()` just means `it is entered when run() is called next`.

Or, if none of that appeals, we should at least document that `run()` should only be called once?

Reply:
Good point.

I change it so that their context is entered in in `__enter__` but additionally, if the context is already active, then `regsiter_api(x)` will also immediately enter the context of `x`. 
Extended the `TestAPI.test_buildroot` test case to check multiple `BuildRoot.run()` invocations work as expected.